text/0000-iri-v181-transaction-and-bundle/0000-iri-v181-transaction-and-bundle.md

```diff
@@ -0,0 +1,980 @@
+    }
+
+    /// Seals the bundle builder and sets the bundle hash on all transaction builders contained in it.
+    ///
+    /// This function calculates the bundle hash from its contained transactions and sets it for all of them, sets
+    /// the transactions' indices, ensures that the M-Bug does not occur, and prevents any extra transactions from being
+    /// pushed into it.
+    ///
+    /// Finalizing a bundle means computing the bundle hash, verifying that it matches the security requirement and setting it
+    /// to all the transactions of the bundle. After finalization, transactions of a bundle are ready to be safely attached to
+    /// the tangle. Finalization is also the place to set the transaction indexes since the bundle essence contains all
+    /// `current_index` and `last_index` and we wouldn't be able to alter them after the bundle hash has been calculated.
+    /// 
+    /// # Normalization
+    ///
+    /// During signing scheme, the share of the private key being leaked is not uniform, introducing a risk
+    /// to leak a critically large part of it. By normalizing the hash, we ensure that exactly half of the private key is
+    /// leaked. The actual normalization algorithm is provided in the signing scheme RFC.
+    /// Useful links: [Addresses and signatures](https://docs.iota.org/docs/dev-essentials/0.1/concepts/addresses-and-signatures),
+    /// [Why is the bundle hash normalized?](https://iota.stackexchange.com/questions/1588/why-is-the-bundle-hash-normalized).
+    /// 
+    /// # M-Bug
+    ///
+    /// Due to the implementation of the signature scheme, the normalized bundle hash can't contain a `M` (or `13`)
+    /// because it could expose a significant part of the private key, making it easier for attackers to forge signatures.
+    /// The bundle hash is then repetitively generated with a slight modification until its normalization doesn't contain a `M`.
+    /// This modification is usually operated by incrementing the `obsolete_tag` of the first transaction of the bundle since
+    /// this field is not being used for any other reason. Useful links:
+    /// [Why is the normalized hash considered insecure when containing the char 'M'](https://iota.stackexchange.com/questions/1241/why-is-the-normalized-hash-considered-insecure-when-containing-the-char-m).
+    /// 
+    /// *Since signature size depends on the security level, a single signature can spread out to up to 3 transactions.
+    /// `inputs` is an object that contains all unused addresses of a seed with a sufficient balance.*
+    /// 
+    /// TODO: This function relies on some `Sponge` trait, which is not yet defined. The code below lays out
+    ///       how the sponge is thought to be used.
+    pub fn seal<S: Sponge<Output=BundleHash>>(self, sponge: S) -> Result<SealedOutgoingBundleBuilder, OutgoingBundleBuilderError> {
+        unimplemented!()
+
+        let mut current_index = 0;
+
+        for transaction_draft in bundle_builder {
```

Comment:
We use both `transaction_draft` and `transaction_draft` in this RFC, can we harmonize ?

Reply:
Done